src/shard.rs

```diff
@@ -332,6 +332,32 @@ impl<
         })
     }
 
+    pub fn retain<F>(&mut self, f: F)
+    where
+        F: Fn(&Key, &Val) -> bool,
+    {
+        let retained_tokens = self
+            .map
+            .iter()
+            .filter_map(|&idx| match self.entries.get(idx) {
+                Some((entry, _idx)) => match entry {
+                    Entry::Resident(r) => {
+                        if f(&r.key, &r.value) {
+                            let hash = self.hash(&r.key);
```

Comment:
One thing I am confused is: why many apis need to pass `hash` from outside sync/unsync? Can I use `self.hash()` to calculate the hash here (since this API doesn't have a `key` to calculate in outside) ?

Reply:
Hash usually appears in the arguments, so the Sync cache can compute hash to select the shard and pass that to the Shard, avoiding a second hash calculation.